torchrl/envs/model_based.py

```diff
@@ -0,0 +1,106 @@
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from __future__ import annotations
+
+import abc
+from typing import Optional, Union, List
+
+import numpy as np
+import torch
+import torch.nn as nn
+
+from torchrl.data import TensorDict
+from torchrl.modules.tensordict_module.world_models import WorldModelWrapper
+from ..data.utils import DEVICE_TYPING
+from ..modules.tensordict_module import TensorDictModule, TensorDictSequence
+from .common import EnvBase
+
+
+class ModelBasedEnv(EnvBase, metaclass=abc.ABCMeta):
+    """
+    Basic environnement for Model Based RL algorithms.
+
+    This class is a wrapper around the model of the MBRL algorithm.
+    This class is meant to give an env framework to a world model and a reward model.
+    It is meant to behave as a classical environment.
+
+    Properties:
+        - observation_spec (CompositeSpec): sampling spec of the observations;
+        - action_spec (TensorSpec): sampling spec of the actions;
+        - input_spec (CompositeSpec): sampling spec of the actions and/or other inputs;
+        - reward_spec (TensorSpec): sampling spec of the rewards;
+        - batch_size (torch.Size): number of environments contained in the instance;
+        - device (torch.device): device where the env input and output are expected to live
+        - is_done (torch.Tensor): boolean value(s) indicating if the environment has reached a done state since the
+            last reset
+
+    Args:
+        world_model (nn.Module): model that generates world states and its corresponding rewards;
+        device (torch.device): device where the env input and output are expected to live
+        dtype (torch.dtype): dtype of the env input and output
+        batch_size (torch.Size): number of environments contained in the instance
+
+    Methods:
+        step (TensorDict -> TensorDict): step in the environment
+        reset (TensorDict, optional -> TensorDict): reset the environment
+        set_seed (int -> int): sets the seed of the environment
+        rand_step (TensorDict, optional -> TensorDict): random step given the action spec
+        rollout (Callable, ... -> TensorDict): executes a rollout in the environment with the given policy (or random
+            steps if no policy is provided)
+
+    """
+
+    def __init__(
+        self,
+        world_model: Union[nn.Module, List[TensorDictModule], TensorDictSequence],
+        params: Optional[List[torch.Tensor]] = None,
+        buffers: Optional[List[torch.Tensor]] = None,
+        device: DEVICE_TYPING = "cpu",
+        dtype: Optional[Union[torch.dtype, np.dtype]] = None,
+        batch_size: Optional[torch.Size] = None,
+    ):
+        super(ModelBasedEnv, self).__init__(
+            device=device, dtype=dtype, batch_size=batch_size
+        )
+        self.world_model = world_model
+        self.world_model_params = params
```

Comment:
here we're doing some weird stuff: we have a functional module that is not functional
it's true that we do the same for the losses though
I wonder if we could pass params and buffers in the step and reset? Maybe not super easy to do as `step` calls `_step`, so we'd need to refactor `step` too to allow it to pass kwargs to `_step`...